drivers/ethernet/eth_stm32_hal.c

```diff
@@ -221,18 +221,32 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	tx_config.Length = total_len;
 	tx_config.TxBuffer = tx_buffer_def;
 
-	hal_ret = HAL_ETH_Transmit(heth, &tx_config, ETH_DMA_TRANSMIT_TIMEOUT);
+	/* tx_buffer is allocated on function stack, we need to wait for the transfer to complete */
+	/* So it is not freed before the interrupt happens */
+	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_Transmit: failed!");
+		res = -EIO;
+		goto error;
+	}
+
+	/* Wait for end of TX buffer transmission */
```

Comment:
Just a question - would it make a sense to allocate the TX buffer(s) outside the stack, use sequentionally all available descriptors and release the memory inside the interrupt, instead of waiting here?

Reply:
> Just a question - would it make a sense to allocate the TX buffer(s) outside the stack, use sequentionally all available descriptors and release the memory inside the interrupt, instead of waiting here?

Thanks for good suggestion. But, I think it may not necessarily bring big performance gain, since:

1. Memory allocation/deallocation especially in interrupt may be performance consuming overhead.
2. In embedded systems it may be very helpful knowing the memory usage of whole system in advance, especially for the buffers, which are written/read permanently.

What we can do: Allocate doubled buffers of each type needed for one tx/rx operation. Thus we can perform eth action on say buffers set A while filling buffer set B. Then just switch them by setting correct pointer. Unfortunately it will increase memory usage, but may increase performance only a little.